plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -596,27 +601,158 @@ public boolean rename(Path src, Path dst)
     public boolean delete(Path path, boolean recursive)
             throws IOException
     {
-        try {
-            if (!directory(path)) {
-                return deleteObject(keyFromPath(path));
+        String bucketName = getBucketName(uri);
+        String key = keyFromPath(path);
+        ObjectMetadata s3ObjectMetadata = getS3ObjectMetadata(path, bucketName, key);
+        ObjectMetadata s3DirectoryMetadata = null;
+        if (!key.isEmpty()) {
+            s3DirectoryMetadata = getS3ObjectMetadata(path, bucketName, key + PATH_SEPARATOR);
```

Comment:
This logic seems off for two reasons:
- non-recursive delete call shouldn't need this extra `getObjectMetadata`
- recursive delete to `_$folder$` suffixed entries should have `stripDirectorySuffix` called on the, no?

Reply:
> non-recursive delete call shouldn't need this extra getObjectMetadata

This seems is a particularity of AWS S3 (not present in MinIO).
`io.trino.plugin.hive.s3.TestTrinoS3FileSystemAwsS3#testDeleteNonRecursivelyPrefixCorrespondingToNamesakeObjectAndEmptyDirectory` showcases that for deleting a key corresponding to an object and a namesake  directory we need two calls.

Reply:
Yeah, this would be one of the somewhat incoherent case where a path could be simultaneously a "file" and "directory" based on how we're trying to map the semantics. For example:
- Objects `s3://bucket/object` and `s3://bucket/object/` both exist.  Is "object" a file? A folder? The filesystem logic allows both to be true depending on the operation in question. `FileSystem.getLocatedFileStatus` and `FileSystem.listLocatedStatus` would both succeed with different results (unless "object" is marked `application/x-directory` in which case they would both call it a directory). 
- Objects `s3://bucket/prefix.txt` and `s3://bucket/prefix.txt/suffix.txt` both exist. Is `prefix.txt` a file? Folder? Same issue, could be either depending on the operation and whether `prefix.txt` is marked with `application/x-directory`.

I'm not sure how much of the potential craziness we really want to handle here if no S3 file systems in the wild actually create objects with these patterns.

Here's perhaps a helpful assumping: the caller has some intention about what is being deleted. Either they expect to delete a "file" or expect to delete a "directory", and will almost certainly pass `recursive = true` for the directory case and `false` for the file case. I don't expect it to be common for a directory deletion to pass "recursive: false" (although we do have to handle that scenario when the object in question is unambiguously a directory and uphold the "must be non-empty" contract). On that assumption, I think we can avoid this extra metadata lookup when `recursive == false` and move it inside the branch handling recursive deletes, and ignore the schizophrenic `s3://bucket/key` and `s3://bucket/key/` co-existing scenario for simple deletions.